rdtools/soiling.py

```diff
@@ -701,3 +704,176 @@ def soiling_srr(energy_normalized_daily, insolation_daily, reps=1000,
+
+    confidence_level : float, default 68.2
+        The size of the confidence interval to use in determining the upper and lower
+        quantiles reported in the returned DataFrame. (The median is always included in
+        the result.)
+
+    Returns
+    -------
+    pd.DataFrame
+        DataFrame describing monthly soiling rates.
+        Columns:
+
+        * `month` - Integer month, January (1) to December (12)
+        * `soiling_rate_median` - The median soiling rate for the month over the
+          entire dataset. Negative indicates soiling is occurring
+        * `soiling_rate_low` - The lower edge of the confidence interval for
+          the monthly soiling rate.
+        * `soiling_rate_high` - The upper edge of the confidence interval
+          for the monthly soiling rate.
+        * `interval_count` - The number of soiling intervals contributing to the
+          monthly calculation. If this number is low, the
+          confidence interval is likely to underestimate the
+          true uncertainty.
+    '''
+
+    # filter to intervals of interest
+    rel_error = 100 * abs((soiling_interval_summary['slope_high'] -
+                           soiling_interval_summary['slope_low']) / soiling_interval_summary['slope'])
+    intervals = soiling_interval_summary[(soiling_interval_summary['length'] >= min_length) &
+                                         (soiling_interval_summary['valid']) &
+                                         (rel_error <= max_relative_slope_error)
+                                         ].copy()
+
+    # count the overlap of each interval with each month
+    month_counts = []
+    for _, row in intervals.iterrows():
+        month_counts.append(_count_month_days(row['start'], row['end']))
+
+    # divy up the monte carlo reps based on overlap
+    for month in range(1, 13):
+        days_in_month = np.array([x[month] for x in month_counts])
+        if days_in_month.sum() > 0:
+            intervals[f'samples_for_month_{month}'] = np.ceil(
+                days_in_month / days_in_month.sum() * reps)
+        else:
+            intervals[f'samples_for_month_{month}'] = 0
+        intervals[f'samples_for_month_{month}'] = intervals[f'samples_for_month_{month}'].astype(int)
+
+    # perform the monte carlo month by month
+    ci_quantiles = [0.5 - confidence_level/2/100, 0.5 + confidence_level/2/100]
+    monthly_rate_data = []
+    relevant_interval_count = []
+    for month in range(1, 13):
+        rates = []
+        sample_col = f'samples_for_month_{month}'
+
+        relevant_intervals = intervals[intervals[sample_col] > 0]
+        for _, row in relevant_intervals.iterrows():
+            rates.append(np.random.uniform(
+                row['slope_low'], row['slope_high'], row[sample_col]))
+
+        rates = [x for sublist in rates for x in sublist]
+
+        if rates:
+            monthly_rate_data.append(np.quantile(rates, [0.5, ci_quantiles[0], ci_quantiles[1]]))
```

Comment:
A sum of N uniform distributions is called the Irwin-Hall distribution F_N. It's quantile function (inverse of the CDF) is not expressible algebraically (it's the inverse of a summation that looks like the binomial formula). I'd say sampling is the practical choice for evaluating quantiles in the simple case where each component distribution is U[0, 1], let alone the complicated case with scaled, shifted and unequally weighted components.

Thanks for the lunchtime math distraction :)

Reply:
Ha!  I was looking at Irwin-Hall last night.  I eventually decided that the situation here is actually the union of uniform distributions, not the sum.  Maybe "mixture" is the correct term instead of "union".  By the time I realized this, it was bedtime and I didn't look into it any further.   In case you are interested in additional distraction, here is the formula for the PDF of the sum of non-identical uniform distributions for N=2 (surely not novel, I just did it as an exercise): https://gist.github.com/kanderso-nrel/acd2f742bba81fb75139d9176d0364d8

Reply:
I interpreted the code as calculating the quantiles of a union of samples from the underlying distributions. In my mind, the CDF underlying the union of samples is the mixture distribution and would look like a weighted sum of the CDFs of each component. I viewed the random variable X as drawn from the mixture distribution which would look like the binomial formula, rather than X = Sum of random variables from each component distribution. I could have that wrong.